### DISCUSSION TITLE
Invitation to collaborate on qemu-3dfx port

### DISCUSSION THREAD

**Discussion by hughobrien** [hughobrien/qemu-3dfx#7] [Announcements]
Hello!

There's a number of forks of this project, it would be excellent if we could find a way to collaborate to pool our efforts.

Has anyone tried building the DOS wrappers? Or adding some automated tests?

**Comment by hughobrien** [comment]
I've pinged five of you who all seem to have made some attempts to fork/improve/build qemu-3dfx. I'm still working through some problems with the overlay myself!

Please introduce yourself, e.g. where are you based, what led you down this rabbit hole, etc. :)

I'm from Ireland, living in the US. I wanted to play an older game that had poor wine support and was disappointed at how hard GPU support is on older OSs. Dream would be to write a virgl driver for windows 98 so it can be all fully native.

I wanted to build my own binaries as I use NixOS and everything is bloody difficult unless you build it yourself. [Here's my changes](https://github.com/hughobrien/nixpkgs/blob/qemu-3dfx/pkgs/applications/virtualization/qemu/default.nix#L146) to nixpkgs if you're interested. I still need to bundle openglide.

Here are our various independent commits to browse:

* [AleksandarBayrev](https://github.com/AleksandarBayrev/qemu-3dfx/commits/master)
* [Hatsum](https://github.com/Hatsum/qemu-win/commits/master)
* [GiftShower](https://github.com/GiftShower/qemu-3dfx/commits/master)
* [superheher](https://github.com/superheher/qemu-3dfx/commits/master)
* [kghorvath](https://github.com/kghorvath/qemu-3dfx/releases)
* [hughobrien](https://github.com/hughobrien/qemu-3dfx/commits/master)

What are some suggestions for features we can add?
* watcom2 support
* simpler build process
* CI
* Upstream some changes to kjliew or qemu

There's a new [softgpu](https://github.com/JHRobotics/softgpu) project also that is quite exciting.

Look forward to meeting you.

**Comment by casasfernando** [comment]
Hi @hughobrien,

I started to look into this project a few months ago because I wanted to play the old Need For Speed games from the end of the 90s using the original CDs (without having to revert to fan patches, etc. to make them work) on modern systems.
I only work on Linux and last January I wrote a script to build qemu, openglide host libraries and all the wrappers successfully using a Docker container.
Glide games are working fine in general (I can't remember if I got to test the DOS wrapper tough), with great performance, but on the other hand I never managed to make DirectX or OpenGL games work.
I'm pretty sure I'm missing pieces on that front like wined3d libraries among other things but I didn't look further into it since the games I wanted to play were mostly Glide.

Since January I didn't touch this again due to lack of spare time and lost of interest, but currently I'm thinking on trying to implement USB gamepad emulation through SDL2. Some QEMU based projects already did something similar (e.g.: XEMU) plus there is patch in reddit (referenced by kjliew in the wiki) that can work as guidance/starting point. It's a bit frustrating having to waste time on this feature since kjliew already implemented it but he is not releasing that part of the code.

If interested in my build script I don't mind sharing it. I didn't published it anywhere yet because it is a work in progress (e.g.: didn't look into making watcom installation non-interactive-).

Thanks for the initiative and would love to see this getting traction and more people to collaborate and contribute.

**Comment by casasfernando** [comment]
@hughobrien did you try to run any OpenGL or D3D game?
I can't make it work. I'm still stuck with this: https://github.com/kjliew/qemu-3dfx/issues/75
Already verified that GTK support is not built in QEMU and also using `-display sdl` when running the VM.

Any ideas?

**Reply by casasfernando** [reply]
Nevermind I already found the cause of the problem and solved it. I was missing a symlink in my system for libGL.so.

**Comment by casasfernando** [comment]
Going back to the original discussion, I cleaned up my build script a lot, now it doesn't need any user interaction at all.
Everything runs on a Debian 12 container, building qemu-3dfx, the wrappers, openglide and packing everything up to an output directory for convenience.
I'm working now on passing runtime variables to the container so one can decide which version of qemu to build, which wrappers, etc.
Once I get that done I will start looking into the SDL2 gamepad emulation.

**Comment by hughobrien** [comment]
Hey @casasfernando  - apologies I was away from things for a few weeks.

I've been using Nix to build things but ultimately the steps should be similar, if you'd like to could you share your build files as a gist and we can compare notes?

Here's mine: 
https://github.com/hughobrien/nixpkgs/blob/be976f0e2bb5d7b407aa6d9dc8c520280fe0847b/pkgs/applications/virtualization/qemu/default.nix#L147

https://github.com/hughobrien/nixpkgs/blob/be976f0e2bb5d7b407aa6d9dc8c520280fe0847b/pkgs/applications/virtualization/qemu-3dfx-wrappers/default.nix
